zebra-state/src/service/check/anchors.rs

```diff
@@ -3,28 +3,29 @@
 
 use std::{collections::HashMap, sync::Arc};
 
-use zebra_chain::{
-    block::{Block, Height},
-    sprout, transaction,
-};
+use zebra_chain::{block::Height, sprout, transaction::Transaction};
 
 use crate::{
     service::{finalized_state::ZebraDb, non_finalized_state::Chain},
-    PreparedBlock, ValidateContextError,
+    ValidateContextError,
 };
 
 /// Checks the final Sapling and Orchard anchors specified by transactions in this
-/// `prepared` block.
+/// `transactions` iter.
 ///
 /// This method checks for anchors computed from the final treestate of each block in
 /// the `parent_chain` or `finalized_state`.
-#[tracing::instrument(skip(finalized_state, parent_chain, prepared))]
-pub(crate) fn sapling_orchard_anchors_refer_to_final_treestates(
+pub(crate) fn sapling_orchard_anchors_refer_to_final_treestates<
+    'a,
+    TransactionsIter: Iterator<Item = &'a Arc<Transaction>>,
+>(
     finalized_state: &ZebraDb,
-    parent_chain: &Chain,
-    prepared: &PreparedBlock,
+    parent_chain: Option<&Arc<Chain>>,
+    transactions: TransactionsIter,
+    height: Option<Height>,
 ) -> Result<(), ValidateContextError> {
-    for (tx_index_in_block, transaction) in prepared.block.transactions.iter().enumerate() {
+    for (tx_index_in_block, transaction) in transactions.enumerate() {
+        let tx_index_in_block = height.map(|_| tx_index_in_block);
```

Comment:
This code is hard to understand - it was hard for me to see that `tx_index_in_block` is `None` when height is `None`. And there is no comment explaining why we do this.

Instead, we could move the loop to the block wrapper function, leaving this function as just the loop body. Then the mempool transaction wrapper can just call it once.

This comment applies to all the modified functions in this module.

Reply:
Sorry about that, updated in [`00c2a4b`](https://github.com/ZcashFoundation/zebra/pull/5716/commits/00c2a4bb1cb67b1ebaa00ec615ddb3257b6c75da) and [`55232d1`](https://github.com/ZcashFoundation/zebra/pull/5716/commits/55232d14135a0e82ae17d1c150223dbf4ded989d).